### PR TITLE
Do not fail when storing to the build cache

### PR DESCRIPTION
**Pull request by wolfs**
We handle errors when storing to the build cache gracefully and disable the build cache instead of failing the build.

### CODE REVIEW

File: subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    @Rule
+    HttpBuildCache httpBuildCache = new HttpBuildCache(testDirectoryProvider)
+
+    def setup() {
+        buildFile << """   
+            import org.gradle.api.*
+            apply plugin: 'base'
+
+            @CacheableTask
+            class CustomTask extends DefaultTask {
```
Comment (big-guy) [inline subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy]: We've hit a bit of a critical mass of test `@CacheableTask` tasks... I wonder if we should pull these up into a fixture so we can just do something like `buildFile << CacheableTasks.simple()` or `buildFile << CacheableTasks.generateFile(size)`.

WDYT? (Not saying you should do it)

File: subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        then:
+        output ==~ /(?s).*org\.gradle\.api\.GradleException: Could not pack property 'outputFile': (Broken pipe|Connection reset).*/
+    }
+
+    private void startServer() {
+        httpBuildCache.start()
+        settingsFile << useHttpBuildCache(httpBuildCache.uri)
+    }
+
+    private static String useHttpBuildCache(URI uri) {
```
Comment (big-guy) [inline subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy]: Should we move this into a `RemoteBuildCacheFixture`?

File: subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                }
+                remote(org.gradle.caching.http.HttpBuildCache) {
+                    url = "${uri}/"   
+                    push = true
+                }
+            }
+        """
+    }
+
+    def withHttpBuildCache() {
```
Comment (big-guy) [inline subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy]: I think we should just stick with `withBuildCache()` unless there's some ambiguity in the integ tests

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy [hunk 4]
```diff
@@ -107,20 +114,45 @@ class CachedTaskExecutionErrorHandlingIntegrationTest extends AbstractIntegratio
...
-        output.count("The remote build cache is now disabled because 3 errors were encountered") == 1
-        output.count("The remote build cache was disabled during the build after encountering 3 errors.") == 1
+        output.count("The remote build cache is now disabled because 3 recoverable errors were encountered.") == 1
+        output.count("The remote build cache was disabled during the build because 3 recoverable errors were encountered.") == 1
+
+        expect:
+        succeeds "clean"
+
+        when:
+        succeeds "assemble"
```
Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy]: Could we put this together (`succeeds "clean", "assemble"`)? and add a comment like, `// build cache should be still enabled during next build`

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy [hunk 5]
```diff
@@ -107,20 +114,45 @@ class CachedTaskExecutionErrorHandlingIntegrationTest extends AbstractIntegratio
...
+        then:
+        !output.contains("The remote build cache is now disabled")
+        !output.contains("The remote build cache was disabled during the build")
```
Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy]: and should we check if any of the tasks came from the cache?

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy [hunk 6]
```diff
@@ -107,20 +114,45 @@ class CachedTaskExecutionErrorHandlingIntegrationTest extends AbstractIntegratio
...
+    }
+
+    def "cache switches off after first non-recoverable error for the current build"() {
+        // We require a distribution here so that we can capture
+        // the output produced after the build has finished
+        executer.requireGradleDistribution()
+        executer.withBuildCacheEnabled()
+        executer.withStackTraceChecksDisabled()
```
Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy]: Maybe extract this into a separate method?

File: subprojects/core/src/main/java/org/gradle/caching/internal/ShortCircuitingErrorHandlerBuildCacheServiceDecorator.java [hunk 7]
```diff
@@ -64,29 +71,38 @@ public boolean load(BuildCacheKey key, BuildCacheEntryReader reader) {
...
-                LOGGER.warn("The {} build cache is now disabled because {} errors were encountered", getRole(), maxErrorCount);
-            }
+            disableBuildCache("{} recoverable errors were encountered.");
+        }
+    }
+
+    private void disableBuildCache(String message) {
+        if (enabled.compareAndSet(true, false)) {
+            disableMessage = message;
+            LOGGER.warn("The {} build cache is now disabled because " + message, getRole(), maxErrorCount);
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/caching/internal/ShortCircuitingErrorHandlerBuildCacheServiceDecorator.java]: Should we include the build cache key here too?

Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/caching/internal/ShortCircuitingErrorHandlerBuildCacheServiceDecorator.java]: Just before printing this message we logged the warning including the build cache key. I was thinking that we could print some information on why (exception/build cache key) we disabled the cache here and at the end of the build. We would then need to remember this information in the class, too (not only the message as we do currently).
WDYT?

File: subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [inline subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy]: Sounds like a good idea. I think we should open a new issue for this.

File: subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy (hunk 2 above)
Comment (wolfs) [inline subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy]: fixed

File: subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [inline subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy]: fixed

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy (hunk 4 above)
Comment (wolfs) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy]: fixed

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy (hunk 5 above)
Comment (wolfs) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy]: fixed

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy (hunk 6 above)
Comment (wolfs) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionErrorHandlingIntegrationTest.groovy]: fixed

File: subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy (hunk 1 above)
Comment (big-guy) [inline subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/HttpBuildCacheServiceErrorHandlingIntegrationTest.groovy]: OK, done https://github.com/gradle/task-output-cache/issues/616

File: subprojects/core/src/main/java/org/gradle/caching/internal/ShortCircuitingErrorHandlerBuildCacheServiceDecorator.java (hunk 7 above)
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/caching/internal/ShortCircuitingErrorHandlerBuildCacheServiceDecorator.java]: OK, I think it's OK as-is.

File: subprojects/core/src/main/java/org/gradle/caching/internal/ShortCircuitingErrorHandlerBuildCacheServiceDecorator.java [hunk 8]
```diff
@@ -64,29 +71,38 @@ public boolean load(BuildCacheKey key, BuildCacheEntryReader reader) {
...
     }
 
     @Override
     public void close() throws IOException {
+        LOGGER.debug("Closing {} build cache", getRole());
         if (!enabled.get()) {
-            LOGGER.warn("The {} build cache was disabled during the build after encountering {} errors.",
-                getRole(), maxErrorCount
-            );
+            LOGGER.warn("The {} build cache was disabled during the build because " + disableMessage, getRole(), maxErrorCount);
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/caching/internal/ShortCircuitingErrorHandlerBuildCacheServiceDecorator.java]: looks like this is missing a set of `{}`?

File: subprojects/core/src/main/java/org/gradle/caching/internal/ShortCircuitingErrorHandlerBuildCacheServiceDecorator.java [hunk 9]
```diff
@@ -64,29 +71,38 @@ public boolean load(BuildCacheKey key, BuildCacheEntryReader reader) {
...
         }
         super.close();
     }
 
     private void recordFailure() {
         if (remainingErrorCount.decrementAndGet() <= 0) {
-            if (enabled.compareAndSet(true, false)) {
-                LOGGER.warn("The {} build cache is now disabled because {} errors were encountered", getRole(), maxErrorCount);
-            }
+            disableBuildCache("{} recoverable errors were encountered.");
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/caching/internal/ShortCircuitingErrorHandlerBuildCacheServiceDecorator.java]: I think we should just format this as we want it vs using SLF4J substitutions.  It makes the `LOGGER.warn` look wrong below (unless you know `message` has a `{}`).

**Comment by big-guy** [comment]
@wolfs this is pretty good.  Feel free to rebase/merge and follow up on my other comments afterwards.

**Pull request merged**